### PR TITLE
feat: Cell references

### PR DESCRIPTION
**Pull request by richburdon**
- Resolves https://github.com/dxos/dxos/issues/7984
- Resolves https://github.com/dxos/dxos/issues/7968

## Tasks
- select items by reference prop
- backspace deletes cell value
- callback to create new row

## Notes
- Table imperative handle to avoid setters on model.
- Tests are currently disabled since plugin-table has deps on other plugins. Factor out to react-ui-table (and deprecate existing).
- added typedefs, generics, push down refs, etc. to reduce deps.
- standardize column => field

## Demo
https://github.com/user-attachments/assets/163f6d2f-f92a-4ff2-8b4f-96fe68d27905

### CODE REVIEW

File: packages/ui/react-ui-grid/src/CellEditor/CellEditor.tsx [hunk 1]
```diff
@@ -174,7 +175,10 @@ export const CellEditor = ({
     <div
       ref={parentRef}
       className={editorVariants[variant].root}
-      style={box}
+      style={{
+        ...box,
+        ...{ '--dx-gridCellWidth': box ? `${box.inlineSize - 1}px` : undefined },
```
Comment (thure) [inline packages/ui/react-ui-grid/src/CellEditor/CellEditor.tsx:180]: Why -1?

File: packages/ui/react-ui-grid/src/Grid/Grid.tsx [hunk 2]
```diff
@@ -109,7 +112,9 @@ const GridContent = forwardRef<NaturalDxGrid, GridScopedProps<GridContentProps>>
   const { id, editing, setEditBox, setEditing } = useGridContext(GRID_CONTENT_NAME, props.__gridScope);
   const [dxGrid, setDxGridInternal] = useState<NaturalDxGrid | null>(null);
 
-  // NOTE(thure): using `useState` instead of `useRef` works with refs provided by `@lit/react` and gives us a reliable dependency for `useEffect` whereas `useLayoutEffect` does not guarantee the element will be defined.
+  // TODO(burdon): Can we use useImperativeHandle here?
```
Comment (thure) [inline packages/ui/react-ui-grid/src/Grid/Grid.tsx:115]: I believe I demonstrated that `useImperativeHandle` causes a regression to the original issue; consumers still need to use `useState` to get a reliable effect dependency.

**Review by thure — APPROVED** [review]
The grid bits look fine to me, just left a question and a remark. The other business superficially looks fine to me, though I'm less familiar and would defer to someone more familiar.

File: packages/ui/react-ui-card/project.json [hunk 3]
```diff
@@ -14,7 +14,8 @@
           "{projectRoot}/src/index.ts"
         ],
         "platforms": [
-          "browser"
+          "browser",
+          "node"
         ]
```
Comment (wittjosiah) [inline packages/ui/react-ui-card/project.json:19]: this is the default, just remove

File: packages/ui/react-ui-data/project.json [hunk 4]
```diff
@@ -14,7 +14,8 @@
           "{projectRoot}/src/index.ts"
         ],
         "platforms": [
-          "browser"
+          "browser",
+          "node"
         ]
```
Comment (wittjosiah) [inline packages/ui/react-ui-data/project.json:19]: remove

File: packages/ui/react-ui-list/project.json [hunk 5]
```diff
@@ -14,7 +14,8 @@
           "{projectRoot}/src/index.ts"
         ],
         "platforms": [
-          "browser"
+          "browser",
+          "node"
         ]
       }
```
Comment (wittjosiah) [inline packages/ui/react-ui-list/project.json:20]: remove

File: packages/sdk/schema/src/types/view.ts [hunk 6]
```diff
@@ -8,6 +8,9 @@ import { create, JsonProp, JsonSchemaType, QueryType, type ReactiveObject, S, Ty
  * Stored field metadata (e.g., for UX).
  */
 export const FieldSchema = S.Struct({
+  // TODO(burdon): We should add a fieldId distinct from the property since in principle we might have
```
Comment (wittjosiah) [inline packages/sdk/schema/src/types/view.ts:11]: could the fieldId just be the property & format concatenated together?

File: packages/plugins/plugin-table/src/components/TableContainer.tsx [hunk 7]
```diff
@@ -39,7 +39,14 @@ const TableContainer = ({ role, table }: LayoutContainerProps<{ table: TableType
   const queriedObjects = useQuery(space, schema ? Filter.schema(schema) : () => false, undefined, [schema]);
   const filteredObjects = useGlobalFilteredObjects(queriedObjects);
 
-  const handleDeleteRow = useCallback((row: any) => space?.db.remove(row), [space]);
+  // TODO(burdon): Undo action.
```
Comment (wittjosiah) [inline packages/plugins/plugin-table/src/components/TableContainer.tsx:42]: use space plugin intent instead of directly removing, it has undo

**Review by wittjosiah — APPROVED** [review]
How does reference validation work? If it's being done via typeahead rather than a select input, what happens if you stop typing before there's a valid value? Does the cell get cleared on blur or set with an invalid state somehow?

File: packages/ui/react-ui-grid/src/CellEditor/CellEditor.tsx (hunk 1 above)
Comment (richburdon) [reply packages/ui/react-ui-grid/src/CellEditor/CellEditor.tsx]: removed.

File: packages/ui/react-ui-grid/src/Grid/Grid.tsx (hunk 2 above)
Comment (richburdon) [reply packages/ui/react-ui-grid/src/Grid/Grid.tsx:115]: let's discuss today.

**Comment by richburdon** [comment]
> How does reference validation work? If it's being done via typeahead rather than a select input, what happens if you stop typing before there's a valid value? Does the cell get cleared on blur or set with an invalid state somehow?

yes, that's what would happen.
as a next step i'm going to add "create new record" as an option also.
and we need to consider how to nav to the other table.

File: packages/plugins/plugin-table/src/components/TableContainer.tsx (hunk 7 above)
Comment (richburdon) [reply packages/plugins/plugin-table/src/components/TableContainer.tsx:42]: @ZaymonFC

File: packages/ui/react-ui-grid/src/Grid/Grid.tsx (hunk 2 above)
Comment (thure) [reply packages/ui/react-ui-grid/src/Grid/Grid.tsx:115]: Even if this method can be made to be equivalent, what value does it deliver exactly?

File: packages/plugins/plugin-table/src/components/Table/Table.tsx [hunk 8]
```diff
@@ -25,32 +35,98 @@ const blockEndLine =
...
+          if (cell) {
+            dxGrid.updateIfWithinBounds(cell);
+          } else {
+            dxGrid.updateCells(true);
+          }
+        },
+      };
+    },
+    [model, dxGrid],
+  );
```
Comment (ZaymonFC) [inline packages/plugins/plugin-table/src/components/Table/Table.tsx:97]: Nice!

**Review by ZaymonFC — APPROVED** [review]
Looking good!

File: packages/plugins/plugin-table/src/model/table-model.ts [hunk 9]
```diff
@@ -6,108 +6,124 @@ import { computed, effect, signal, type ReadonlySignal } from '@preact/signals-c
...
   pinnedRows?: { top: number[]; bottom: number[] };
   rowSelection?: number[];
-  onDeleteColumn?: (field: string) => void;
-  onDeleteRow?: (row: any) => void;
   onInsertRow?: (index?: number) => void;
+  onDeleteRow?: (row: number, object: T) => void;
+  onDeleteColumn?: (fieldId: string) => void;
   onCellUpdate?: (cell: GridCell) => void;
   onRowOrderChanged?: () => void;
 };
```
[the remark this answers is not in the document]
Comment (richburdon) [reply packages/plugins/plugin-table/src/model/table-model.ts:42]: all of the devtools use tables with definite types, so in that case i think it's useful.

**Pull request merged**